app/async/workers/order_poster.rb

```
@@ -2,30 +2,37 @@ module Workers
   class OrderPoster < AsyncBase
     include Sidekiq::Worker
 
-    sidekiq_options retry: 3
+    sidekiq_options retry: 3,
+                    unique: true,
+                    unique_args: ->(args) { [ args.first ] }
 
     def perform(order_id)
       order = Spree::Order.find(order_id)
       log = File.open("#{Rails.root}/log/#{order.number}_#{Time.now.to_s.gsub(/\s/, "_")}_newgistics_order_poster.log", 'a')
-      log << "#{Time.now}"
-      log << "Posting order #{order.id} with number #{order.number}"
-      document = Spree::Newgistics::DocumentBuilder.build_shipment(order.shipments)
-      response = Spree::Newgistics::HTTPManager.post('/post_shipments.aspx', document)
-      if response.status <= 299
-        errors = Nokogiri::XML(response.body).css('errors').children
-        if !errors.any?
-          log << "NG responded with status #{response.status}, processing order"
-          order.update_attributes({posted_to_newgistics: true, newgistics_status: 'RECEIVED'})
+      log << "#{Time.now} - Posting order #{order.id} with number #{order.number}\n"
+      if !order.posted_to_newgistics
+        log << "Order is being sent to Newgistics\n"
+        document = Spree::Newgistics::DocumentBuilder.build_shipment(order.shipments)
+        response = Spree::Newgistics::HTTPManager.post('/post_shipments.aspx', document)
+        if response.status <= 299
+          errors = Nokogiri::XML(response.body).css('errors').children
+          if !errors.any?
+            log << "NG responded with status #{response.status}, processing order\n"
+            order.update_attributes({posted_to_newgistics: true, newgistics_status: 'RECEIVED'})
+          else
+            log << "NG responded with status #{response.status} with errors:\n"
+            errors.each { |e| log << "#{e}\n" }
```

I think the error handling could be cleansed up a bit. For example, why check for status code if we also check for errors? also, can we consolidate the error handling so that is the same path no matter if the status code is greater or less than 299?

Another thought: it would be nice if we could also raise an error here. Let's talk more on Monday.